libass/ass.c

```diff
@@ -1359,6 +1364,34 @@ void ass_flush_events(ASS_Track *track)
     track->parser_priv->read_order_elems = 0;
 }
 
+void ass_configure_gc(ASS_Track *track, int enabled, long long delay)
+{
+    track->parser_priv->gc_enabled = enabled;
+
+    if (0ll <= delay)
+        track->parser_priv->gc_delay = delay;
+}
+
+#define TS_OUT(ev) (ev.Start + ev.Duration)
+
+// This function is not thread safe
+void ass_prune_events(ASS_Track *track, long long deadline)
+{
+    bool check_readorder = track->parser_priv->check_readorder;
+
+    for (int k = 0, eid = 0; eid < track->n_events; eid = ++k) {
+        while (k < track->n_events && TS_OUT(track->events[k]) < deadline) {
+            if (check_readorder)
+                test_and_write_read_order_bit(track, track->events[k].ReadOrder, false);
+            ass_free_event(track, k++);
+        }
+        if (k > eid) {
+            memmove(&track->events[eid], &track->events[k], sizeof(ASS_Event)*(track->n_events - k));
+            track->n_events -= (k - eid);
```

Comment:
When $N$ non-consecutive events get freed this will move data after the last freed event $N$-times. This can be optimised to only move each remaining chunk once by only moving data after encountering the _next_ chunk of to-be-freed events or the end of the list